### PR TITLE
[2.2] Support for receiving commands to manipulate Rollouts

### PR DESCRIPTION
**Pull request by douglascamata**
## Description
This pull requests add support Rollouts manipulation commands sent by Ambassador's backend through Agent directives and reports back with results of their execution.

Currently 3 commands are supported:

* Pause a rollout.
* Resume rollout.
* Abort rollout.

The report includes:

* The command ID
* Whether it executed successfully or not
* A message (error message in case it didn't execute successfully)

## Related Issues
List related issues.

## Testing

* Automated tests were added to the rollout command execution logic
* Manual integration tests were done using a stub of Ambassador's cloud to ensure the commands will be executed as expected and their results correctly reported.

## Checklist

 - [x] I made sure to update `CHANGELOG.md`.
   
   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations
 
 - [x] This is unlikely to impact how Ambassador performs at scale.
 
   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability
 
 - [x] My change is adequately tested.
 
   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points
 
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

### CODE REVIEW

**Comment by LukeShu** [comment]
I've rebased this on to `lukeshu/upgrade-client-go` and squashed a few commits that should have been squashed.  The diff from a naive `git merge origin/lukeshu/upgrade-client-go` (which I pushed as `lukeshu/merge/dcamata/remote-control-rollouts`, if you want to see it) is as follows:

```patch
diff --git a/builder/Dockerfile b/builder/Dockerfile
index da85aba1a..279e6fc94 100644
--- a/builder/Dockerfile
+++ b/builder/Dockerfile
@@ -79,6 +79,7 @@ ADD pkg pkg
 ADD vendor vendor
 ADD go.mod go.mod
 ADD go.sum go.sum
+
 RUN --mount=type=cache,target=/root/.cache/go-build \
     mkdir -p /go/bin && \
 	time go build -mod=vendor -o /go/bin/ ./cmd/...
diff --git a/go.mod b/go.mod
index cb8d5f737..03fb51bcc 100644
--- a/go.mod
+++ b/go.mod
@@ -48,6 +48,7 @@ go 1.17
 require (
 	github.com/Masterminds/semver v1.5.0 // indirect
 	github.com/Masterminds/sprig v2.22.0+incompatible
+	github.com/argoproj/argo-rollouts v1.0.7
 	github.com/asaskevich/govalidator v0.0.0-20200428143746-21a406dcc535 // indirect
 	github.com/census-instrumentation/opencensus-proto v0.2.1
 	github.com/cncf/udpa/go v0.0.0-20210322005330-6414d713912e
@@ -98,8 +99,6 @@ require (
 	sigs.k8s.io/yaml v1.2.0
 )
 
-require github.com/argoproj/argo-rollouts v1.0.7
-
 require (
 	cloud.google.com/go v0.58.0 // indirect
 	github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78 // indirect
```

**Review by rp4rk — DISMISSED** [review]
LGTM

File: pkg/agent/rollouts.go [hunk 1]
```diff
@@ -12,16 +12,22 @@ import (
...
+	// rolloutActionPause represents the "pause" action on a Rollout
+	rolloutActionPause = rolloutAction("pause")
+	// rolloutActionResume represents the "resume" action on a Rollout
 	rolloutActionResume = rolloutAction("resume")
-	rolloutActionAbort  = rolloutAction("abort")
+	// rolloutActionAbort represents the "abort" action on a Rollout
+	rolloutActionAbort = rolloutAction("abort")
 )
 
+// rolloutsGetterFactory is a factory for creating RolloutsGetter.
```
Comment (kflynn) [inline pkg/agent/rollouts.go:27]: So what's the purpose of a `RolloutsGetter`?

File: docs/releaseNotes.yml [hunk 2]
```diff
@@ -54,6 +54,15 @@ items:
           variable have been removed. Only the Envoy V3 API is supported (this has been the
           default since Emissary-ingress v1.14.0).
 
+      - title: Support received commands to pause, continue and abort a Rollout via Agent directives
+        type: feature
+        body: >-
+          The Emissary agent now receive commands to manipulate Rollouts (pause, continue, and abort are
```
Comment (kflynn) [inline docs/releaseNotes.yml:79]: This should really be "receives". 🙂

File: pkg/agent/directive_handler.go [hunk 3]
```diff
@@ -99,7 +99,10 @@ func (dh *BasicDirectiveHandler) reportCommandResult(ctx context.Context, comman
 		result.Success = false
 		result.Message = cmdError.Error()
 	}
-	err := a.comm.ReportCommandResult(ctx, result)
+	a.ambassadorAPIKeyMutex.Lock()
+	apiKey := a.ambassadorAPIKey
+	a.ambassadorAPIKeyMutex.Unlock()
```
Comment (kflynn) [inline pkg/agent/directive_handler.go:104]: Might be worth a comment that you really don't want to hold this lock during the report-command-result call.

**Review by kflynn — DISMISSED** [review]
Need to handle merge conflicts and tests, but this looks good! 🙂  My comments below are things that don't need to block landing this.

**Review by LukeShu — APPROVED** [review]
Resolve the conflicts, and it looks fine to me.

File: pkg/agent/rollouts.go (hunk 1 above)
Comment (douglascamata) [reply pkg/agent/rollouts.go:27]: The purpose of `RolloutsGetter` is to fetch Rollouts in a namespace inside the cluster.

File: pkg/agent/directive_handler.go (hunk 3 above)
Comment (douglascamata) [reply pkg/agent/directive_handler.go:104]: @kflynn Why? Where would you recommend to use the mutex to access the api key from the agent and send it down to this point?

**Review by LukeShu — APPROVED** [review]

**Pull request merged**